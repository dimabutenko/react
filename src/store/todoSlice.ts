import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Filter } from './filter';
import type { RootState } from '../app/store';
import { ITodo, TodoState } from '../type';

export const initialState: TodoState = {
  todos: [],
  filter: Filter.All,
};

export const todoSlice = createSlice({
  name: 'counter',
  initialState,
  reducers: {
    add(state: TodoState, action: PayloadAction<string>) {
      state.todos.push({
        id: Math.random(),
        name: action.payload,
        completed: false,
      });
    },
    update(state: TodoState, action: PayloadAction<ITodo>) {
      const index = state.todos.map((x) => x.id).indexOf(action.payload.id, 0);
      if (index > -1) {
        state.todos[index].completed = action.payload.completed;
      }
    },
    remove(state: TodoState, action: PayloadAction<number>) {
      const index = state.todos.map((x) => x.id).indexOf(action.payload, 0);
      if (index > -1) {
        state.todos.splice(index, 1);
      }
    },
    updateFilter(state: TodoState, action: PayloadAction<Filter>) {
      state.filter = action.payload;
    },
    clearCompleted(state: TodoState) {
      state.todos = state.todos.filter((x) => !x.completed);
    },
  },
});

export const { add, update, remove, updateFilter, clearCompleted } = todoSlice.actions;
export const todoState = (state: RootState) => state.todoState;
export default todoSlice.reducer;
