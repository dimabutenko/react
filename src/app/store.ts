import { configureStore } from '@reduxjs/toolkit'
import todoReducer, { initialState } from '../store/todoSlice'
import { listenerMiddleware } from '../store/middleware'

const todoState = JSON.parse(localStorage.getItem('todoState') ?? 'null')

export const store = configureStore({
  preloadedState: {
    todoState: todoState === null ? initialState : todoState
  },
  reducer: {
    todoState: todoReducer
  },
  middleware: (getDefaultMiddleware) => [...getDefaultMiddleware(), listenerMiddleware.middleware]
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
