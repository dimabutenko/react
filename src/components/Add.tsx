import styles from './Add.module.scss'
import React, { useRef } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCirclePlus } from '@fortawesome/free-solid-svg-icons/faCirclePlus'
import { useAppDispatch } from '../app/hooks'
import { add } from '../store/todoSlice'

const Add: React.FC = () => {
  const inputRef = useRef<HTMLInputElement>(null)
  const dispatch = useAppDispatch()

  function handleClick (): void {
    if (inputRef.current != null) {
      dispatch(add(inputRef.current.value))
      inputRef.current.value = ''
    }
  }

  function handleKeyDown (event: React.KeyboardEvent<HTMLInputElement>): void {
    if (event.key === 'Enter') {
      handleClick()
    }
  }

  return (
    <div className={`card ${styles['add-card']}`}>
      <input
        type="text"
        ref={inputRef}
        className={styles.input}
        placeholder="Create a new todo..."
        onKeyDown={handleKeyDown}
      />
      <button className={styles.button} onClick={handleClick}>
        <FontAwesomeIcon icon={faCirclePlus} size="2xl" className={styles.icon} />
      </button>
    </div>
  )
}

export default Add
