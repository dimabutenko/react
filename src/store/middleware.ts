import { createListenerMiddleware, isAnyOf } from '@reduxjs/toolkit'
import { todoSlice } from './todoSlice'
import { type RootState } from '../app/store'

export const listenerMiddleware = createListenerMiddleware()
listenerMiddleware.startListening({
  matcher: isAnyOf(
    todoSlice.actions.add,
    todoSlice.actions.remove,
    todoSlice.actions.update,
    todoSlice.actions.clearCompleted
  ),
  effect: (action, listenerApi) => {
    localStorage.setItem(
      'todoState',
      JSON.stringify((listenerApi.getState() as RootState).todoState)
    )
  }
})
