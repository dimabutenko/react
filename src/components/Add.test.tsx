import React from 'react'
import { fireEvent, screen, render } from '@testing-library/react'
import Add from './Add'
import { Provider } from 'react-redux'
import { store } from '../app/store'

describe('Add conpoonent tests', () => {
  test('Should render correctly', () => {
    const { container } = render(
      <Provider store={store}>
        <Add />
      </Provider>
    )
    expect(container).toMatchSnapshot()
  })
  test('Should clean input value when key down event with key Enter fired', () => {
    render(
      <Provider store={store}>
        <Add />
      </Provider>
    )
    const input = screen.getByPlaceholderText<HTMLInputElement>('Create a new todo...')
    input.value = 'New TODO'
    fireEvent.keyDown(input, { key: 'Enter' })

    expect(input.value).toEqual('')
  })

  test('Should nothing to do with input value when key down event fired with different than Enter key', () => {
    render(
      <Provider store={store}>
        <Add />
      </Provider>
    )
    const input = screen.getByPlaceholderText<HTMLInputElement>('Create a new todo...')
    input.value = 'New TODO'
    fireEvent.keyDown(input, { key: 'Esc' })

    expect(input.value).toEqual('New TODO')
  })
})
