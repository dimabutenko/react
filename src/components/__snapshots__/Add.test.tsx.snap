// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add conpoonent tests Should render correctly 1`] = `
<div>
  <div
    class="card add-card"
  >
    <input
      class="input"
      placeholder="Create a new todo..."
      type="text"
    />
    <button
      class="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-plus fa-2xl icon"
        data-icon="circle-plus"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM232 344V280H168c-13.3 0-24-10.7-24-24s10.7-24 24-24h64V168c0-13.3 10.7-24 24-24s24 10.7 24 24v64h64c13.3 0 24 10.7 24 24s-10.7 24-24 24H280v64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;
